text/0058-unseal-spancontext.md

```diff
@@ -0,0 +1,117 @@
+
+* Add custom routing information to the SpanContext more efficiently than by using the string-based mapping in TraceState.
+* Generate Span IDs lazily (e.g., if a Span only has internal parents and children, a mechanism that is more efficient than a combined 128 + 64 bit identifier could be used).
+* Count the number of times a SpanContext was injected to better estimate on the back end how long to wait for (additional) child Spans.
+
+## Explanation
+
+SpanContext remains as a non-abstract class implemented in the API, but they are non-final. Note that the proposition
+
+> SpanContexts are immutable.
+
+only applies to the part of the SpanContext that is publicly exposed on the API level.
+Actual vendor implementations could have additional mutable properties or
+actually have the first access to the `TraceId` property generate the TraceId lazily. Vendors must take care not to break any assumptions by doing that. For example, if a language allows copying a SpanContext, vendors should better put all mutable data behind an immutable reference so that all copies remain in a consistent state.
+
+## Internal details
+
+The changes in the initial summary of this OTEP shall be applied to the specification at
+https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#spancontext.
+
+Language libraries that used a "final" or "sealed" keyword or something similar in their API definition of SpanContext must remove it.
+
+The reference SDK implementation will continue to use the API's SpanContext implementation without deriving from it.
+Additionally, all public APIs of the SpanContext must be overridable
+(they must not be public data fields in languages where these are not overridable,
+languages that require an explicit marker like `virtual` to make an API overridable must add it, etc.).
+
+Existing propagators
+(which should be implemented in the SDK package or a separate package)
+can continue to directly construct a SpanContext.
+API implementations that want to use a custom SpanContext implementation must either
+handle having a plain SpanContext as parent
+and having additional properties stripped upon extracting
+or they provide their own propagator implementations too.
+
+## Trade-offs and mitigations
+
+The decision to have SpanContext as a non-abstract class in the API results from the goal to keep compatibility.
+The cleaner alternative would be to have SpanContext be an interface like all other API classes.
+
+Since the implementation of propagators is on the SDK level, all public APIs could be removed from the SpanContext
+so that it becomes an object that is completely opaque (to everyone but the propagator implementations).
+This would enable even more flexibility or at least spare vendors from adding boilerplate code that, e.g., calculates TraceIds
+if there is no such concept in their own system.
```

Comment:
I wouldn't push for this argument, because that was the approach in OpenTracing and there was a lot of grief from people who simply wanted to load the damned trace id.

Reply:
If you just want something to log, then then you could have string properties for Trace and Span ID without forcing them to have a particular format or be the canonical representation. Even a `ToString` override may be sufficient for that.

Reply:
You need to have guarantees of the length and format in some case, even for logging. I think this was discussed as a requirement for OpenTelemetry from the beginning and cannot be changed.

Reply:
These guarantees (such as ASCII-only, no spaces) could be documented for API-implementations to follow instead of enforced by fixing certain types for them.

But I fully agree that this cannot really be changed at this point (the cost-benefit ratio would be very bad here).